dask/dataframe/core.py

```diff
@@ -1130,21 +1166,33 @@ def _bind_operator_method(cls, name, op):
         def meth(self, other, level=None, fill_value=None, axis=0):
             if not level is None:
                 raise NotImplementedError('level must be None')
-            return map_partitions(op, self.column_info, self, other,
+            return map_partitions(op, self._pd, self, other,
                                   axis=axis, fill_value=fill_value)
         meth.__doc__ = op.__doc__
         bind_method(cls, name, meth)
 
     def apply(self, func, convert_dtype=True, name=no_default, args=(), **kwds):
         """ Parallel version of pandas.Series.apply """
+
         if name is no_default:
-            name = self.name
+            msg = ("name is not specified, inferred from partial data. "
+                   "Please provide name if the result is unecpected.\n"
+                   "  Before: s.apply(func)\n"
+                   "  After:  s.apply(func, name=['x', 'y']) for dataframe result\n"
+                   "  or:     s.apply(func, name='x')        for series result")
+            warnings.warn(msg)
+
+            name = _emulate(pd.Series.apply, self.head(), func,
+                            convert_dtype=convert_dtype,
+                            args=args, **kwds)
```

Comment:
In some cases I think that we *can* get result column names and types from empty data.  

For example I think that all of the functions in your tests in `test_apply_infer_columns` would provide correct column and dtype information when called on empty datasets.

In cases where this doesn't work (when some exception is raised) then yes, perhaps calling the function on `self.head()` is appropriate.